### PR TITLE
More NDIS detail

### PR DESCRIPTION
**Pull request by ami-GS**

### CODE REVIEW

File: time-based-packet-transmission-offload.md [hunk 1]
```diff
@@ -153,6 +171,29 @@ typedef enum _NDIS_TXTIME_SUPPORT_FLAGS {
...
 
+To obtain the `NDIS_TCP_IP_TXTIME_NET_BUFFER_LIST_INFO` structure, a driver should call the `NET_BUFFER_LIST_INFO` macro with an _Id of `TcpIpTxTimeNetBufferListInfo`
+```C
+typedef struct _NDIS_TCP_IP_TXTIME_NET_BUFFER_LIST_INFO {
+  struct {
+    ULONG IsIPv4 : 1;
+    ULONG IsIPv6 : 1;
+    ULONG IsTcp : 1;
+    ULONG IsUdp : 1;
+    ULONG TxTime : 64;
```
Comment (csujedihy) [inline time-based-packet-transmission-offload.md:182]: ULONG 64bit?

Comment (csujedihy) [reply time-based-packet-transmission-offload.md:182]: Also, I don't think we can use more than 32 bits in this struct. Each NBL OOB info field is actually a PVOID, which can only fit 32 bits.

File: time-based-packet-transmission-offload.md [hunk 2]
```diff
@@ -153,6 +171,29 @@ typedef enum _NDIS_TXTIME_SUPPORT_FLAGS {
...
 ```
 
+To obtain the `NDIS_TCP_IP_TXTIME_NET_BUFFER_LIST_INFO` structure, a driver should call the `NET_BUFFER_LIST_INFO` macro with an _Id of `TcpIpTxTimeNetBufferListInfo`
+```C
+typedef struct _NDIS_TCP_IP_TXTIME_NET_BUFFER_LIST_INFO {
+  struct {
+    ULONG IsIPv4 : 1;
+    ULONG IsIPv6 : 1;
+    ULONG IsTcp : 1;
+    ULONG IsUdp : 1;
```
Comment (csujedihy) [inline time-based-packet-transmission-offload.md:181]: Why would the NIC want to know it's a V4 or V6 packet? The same goes for TCP and UDP too.

File: time-based-packet-transmission-offload.md (hunk 1 above)
Comment (mtfriesen) [reply time-based-packet-transmission-offload.md:182]: We should scope this offload to 64-bit systems only if that reduces our OOB complexity. I can't imagine these offloads being relevant for 32 platforms.

Comment (nibanks) [reply time-based-packet-transmission-offload.md:182]: The only 32-bit platforms MSFT ships now are x86 for Client SKU. Server is all 64-bit. I'd be perfectly happy scoping to 64-bit if we need to.

File: time-based-packet-transmission-offload.md (hunk 2 above)
Comment (nibanks) [reply time-based-packet-transmission-offload.md:181]: If we can, I definitely would prefer to eliminate this extra stuff too.

File: time-based-packet-transmission-offload.md [hunk 3]
```diff
@@ -153,6 +171,29 @@ typedef enum _NDIS_TXTIME_SUPPORT_FLAGS {
...
+    ULONG TxTime : 64;
+    ULONG DropIfRate : 1;
```
Comment (csujedihy) [inline time-based-packet-transmission-offload.md:183]: what is `DropIfRate`?

File: time-based-packet-transmission-offload.md (hunk 2 above)
Comment (mtfriesen) [reply time-based-packet-transmission-offload.md:181]: Or can we leverage existing OOBs? E.g. reuse some checksum offload fields? It's crazy that each offload has an "IsIpv6" flag...

File: time-based-packet-transmission-offload.md [hunk 4]
```diff
@@ -153,6 +171,29 @@ typedef enum _NDIS_TXTIME_SUPPORT_FLAGS {
...
+  } Transmit;
+  struct {
+    ULONG TcpTxTimeFailed : 1;
+    ULONG UdpTxTimeFailed : 1;
+  } TransmitComplete;
```
Comment (csujedihy) [inline time-based-packet-transmission-offload.md:188]: Is this really needed? Can we just ask the NIC to set NBL::Status?

Comment (csujedihy) [reply time-based-packet-transmission-offload.md:188]: Even if needed, we should union this struct with Transmit.

File: time-based-packet-transmission-offload.md (hunk 1 above)
Comment (csujedihy) [reply time-based-packet-transmission-offload.md:182]: I am not sure what this TxTime represents. If it takes 64 bits, we still can't fit it in a NBL oob.

File: time-based-packet-transmission-offload.md (hunk 3 above)
Comment (ami-GS) [reply time-based-packet-transmission-offload.md:183]: Drop the packet if the packet arrived after deadline specified by TxTime. (Not sure here is the right place)

File: time-based-packet-transmission-offload.md (hunk 1 above)
Comment (ami-GS) [reply time-based-packet-transmission-offload.md:182]: It is time an associated packet to be sent.

Comment (csujedihy) [reply time-based-packet-transmission-offload.md:182]: I know it's time but what is the reference and unit?

File: time-based-packet-transmission-offload.md (hunk 3 above)
Comment (csujedihy) [reply time-based-packet-transmission-offload.md:183]: We should not drop a packet for this reason. If it's later than the deadline, then we send it immediately. It's costly to recover this packet in TCP and it triggers congestion avoidance for no reason.

Comment (ami-GS) [reply time-based-packet-transmission-offload.md:183]: Linux has the option though (I think, it means hardware has the feature already)

Comment (nibanks) [reply time-based-packet-transmission-offload.md:183]: For real-time applications where they don't care about loss, the drop might be ok? That being said, I kind of agree that this seems unnecessary and over complicates the design. Let's simplify things by removing support for this.

Comment (csujedihy) [reply time-based-packet-transmission-offload.md:183]: Do you mind sharing a link? I am curious to learn why linux has it.

File: time-based-packet-transmission-offload.md (hunk 1 above)
Comment (ami-GS) [reply time-based-packet-transmission-offload.md:182]: I'm investigating how windows get time, but Linux set txtime from userspace as bellow
```C
	struct timespec ts;
	int cnt, err;
	__u64 txtime;

	clock_gettime(clkid, &ts);

	/* Start one to two seconds in the future. */
	ts.tv_sec += 1;
	ts.tv_nsec = 1000000000 - waketx_delay;
	normalize(&ts);

	txtime = ts.tv_sec * 1000000000ULL + ts.tv_nsec;
	txtime += waketx_delay;
```

**Comment by csujedihy** [comment]
Probably out of the scope of this PR but just curious have we discussed how we should handle backpressure notifications for TL and app layer?

File: time-based-packet-transmission-offload.md (hunk 3 above)
Comment (ami-GS) [reply time-based-packet-transmission-offload.md:183]: As far as I undnerstand, the queue seems to be allocated on system/hardware. depends on offloading config (there is no explicit explanation though)

File: time-based-packet-transmission-offload.md [hunk 5]
```diff
@@ -95,16 +98,21 @@ if(setsockopt(sock, SOL_SOCKET, SO_TXTIME_SCHED, (char*)&optVal, optLen) == SOCK
...
 typedef struct _TPTO_ANCILLARY_DATA {
     uint64_t TxTime;
-    uint8_t  DropIfRate
+    uint64_t TimeDelta;
 } TPTO_ANCILLARY_DATA;
 ```
 
 - TxTime to indicate when the associated packet is sent
-- DropIfRate to indicate whether to drop packet if it cannot be transmitted by the given deadline
+- TimeDelta is to spread batching packets
```
Comment (maolson-msft) [inline time-based-packet-transmission-offload.md:110]: "TimeDelta is to spread batching packets"- what does this mean?

Comment (ami-GS) [reply time-based-packet-transmission-offload.md:110]: This is for LSO/USO to transmit segmented packet with interval of TimeDelta

File: time-based-packet-transmission-offload.md [hunk 6]
```diff
@@ -95,16 +98,21 @@ if(setsockopt(sock, SOL_SOCKET, SO_TXTIME_SCHED, (char*)&optVal, optLen) == SOCK
...
+
+
+To avoid costly behavior like packet recovery, just send immediately if the packet's TxTime has passed already. (Do not set DropIfRate flag)
```
[the remark this answers is not in the document]
Comment (ami-GS) [reply time-based-packet-transmission-offload.md:113]: DropIfLate.....🫣
Yes Yi recommended to remove DropIfLate. Just send immediately.
Sorry, I thought DropIfLate (DROP_IF_RATE) flag is common for TPTO context

File: time-based-packet-transmission-offload.md [hunk 7]
```diff
@@ -18,6 +18,7 @@ Because of the time-fidelity requirements of this feature, there will not be sof
   - [Sending Packets](#sending-packets)
 - [TCPIP](#tcpip)
 - [NDIS](#ndis)
+- [Registry value based setting](#registry-value-based-setting)
 - [Appendix](#appendix)
   - [Linix](#linix)
```
Comment (csujedihy) [inline time-based-packet-transmission-offload.md:23]: Linix -> Linux

File: time-based-packet-transmission-offload.md [hunk 8]
```diff
@@ -153,6 +178,24 @@ typedef enum _NDIS_TXTIME_SUPPORT_FLAGS {
...
+```C
+typedef struct _NDIS_TCP_IP_TXTIME_NET_BUFFER_LIST_INFO {
+  struct {
+    ULONG IsIPv4 : 1;
+    ULONG IsIPv6 : 1;
+    ULONG IsTcp : 1;
+    ULONG IsUdp : 1;
+    ULONGLONG TxTime : 64;
+    ULONGLONG TimeDelta : 64;
+  } Transmit;
```
Comment (csujedihy) [inline time-based-packet-transmission-offload.md:191]: Pacing only makes sense for TX. No need to wrap everything in `Transmit.` It's implicit.

Comment (ami-GS) [reply time-based-packet-transmission-offload.md:191]: I'm investigating whether TransmitComplete is required for TPTO....

Comment (ami-GS) [reply time-based-packet-transmission-offload.md]: Only TxTime and TimeDelta should be enough

File: time-based-packet-transmission-offload.md [hunk 9]
```diff
@@ -95,16 +98,21 @@ if(setsockopt(sock, SOL_SOCKET, SO_TXTIME_SCHED, (char*)&optVal, optLen) == SOCK
...
 typedef struct _TPTO_ANCILLARY_DATA {
     uint64_t TxTime;
-    uint8_t  DropIfRate
+    uint64_t TimeDelta;
 } TPTO_ANCILLARY_DATA;
 ```
 
 - TxTime to indicate when the associated packet is sent
-- DropIfRate to indicate whether to drop packet if it cannot be transmitted by the given deadline
+- TimeDelta is an interval to transmit segmented packets by LSO/USO
```
Comment (maolson-msft) [inline time-based-packet-transmission-offload.md:110]: The updated text still doesn't make it clear what this means. You need text like "time by which packets in an LSO send should be spread out". (But also, that's not the plan, right? It seems like small bursts of a few packets should be alright and more efficient.)

Comment (ami-GS) [reply time-based-packet-transmission-offload.md:110]: I and Nick discussed yesterday. It is also mentioned in Issue
https://github.com/microsoft/quic-offloads/issues/49

Comment (nibanks) [reply time-based-packet-transmission-offload.md:111]:
```suggestion
- TimeDelta is an interval, starting at `TxTime`, to transmit individual packets that have been coalesced into a single send (by LSO/USO)
```

**Review by nibanks — APPROVED** [review]

**Pull request merged**